integrations/go/workload-executor.go

```diff
@@ -0,0 +1,309 @@
+	}
+
+	return nil, nil
+}
+
+func executeUpdateOne(coll *mongo.Collection, args bson.Raw) (*mongo.UpdateResult, error) {
+	filter := emptyDoc
+	var update interface{} = emptyDoc
+	var err error
+	opts := options.Update()
+
+	elems, _ := args.Elements()
+	for _, elem := range elems {
+		key := elem.Key()
+		val := elem.Value()
+
+		switch key {
+		case "filter":
+			filter = val.Document()
+		case "update":
+			update, err = createUpdate(val)
+			if err != nil {
+				return nil, err
+			}
+		default:
+			str := fmt.Sprintf("unrecognized updateOne option: %v", key)
+			panic(str)
+		}
+	}
+	if opts.Upsert == nil {
+		opts = opts.SetUpsert(false)
+	}
+
+	return coll.UpdateOne(context.Background(), filter, update, opts)
+}
+
+func verifyInsertOneResult(actualResult *mongo.InsertOneResult, expectedResult interface{}) bool {
+	if expectedResult == nil {
+		return true
+	}
+
+	var expected mongo.InsertOneResult
+	if bson.Unmarshal(expectedResult.(bson.Raw), &expected) != nil {
+		return false
+	}
+
+	expectedID := expected.InsertedID
+	if f, ok := expectedID.(float64); ok && f == math.Floor(f) {
+		expectedID = int32(f)
+	}
+
+	return expectedID == nil || (actualResult != nil && expectedID == actualResult.InsertedID)
+}
+
+func verifyCursorResult(cur *mongo.Cursor, result interface{}) bool {
+	if result == nil {
+		return true
+	}
+
+	if cur == nil {
+		return false
+	}
+
+	defer func() {
+		err := cur.Close(context.Background())
+		if err != nil {
+			panic(err)
+		}
+	}()
+
+	for _, expected := range result.(bson.A) {
+		if !cur.Next(context.Background()) {
+			return false
+		}
+		if !bytes.Equal(expected.(bson.Raw), cur.Current) {
+			return false
+		}
+	}
+
+	if cur.Next(context.Background()) {
+		return false
+	}
+	return cur.Err() == nil
+}
+
+func verifyUpdateResult(res *mongo.UpdateResult, result interface{}) bool {
+	if result == nil {
+		return true
+	}
+
+	var expected struct {
+		MatchedCount  int64 `bson:"matchedCount"`
+		ModifiedCount int64 `bson:"modifiedCount"`
+		UpsertedCount int64 `bson:"upsertedCount"`
+	}
+	err := bson.Unmarshal(result.(bson.Raw), &expected)
+	if err != nil {
+		return false
+	}
+
+	if expected.MatchedCount != res.MatchedCount {
+		return false
+	}
+	if expected.ModifiedCount != res.ModifiedCount {
+		return false
+	}
+
+	actualUpsertedCount := int64(0)
+	if res.UpsertedID != nil {
+		actualUpsertedCount = 1
+	}
+	return expected.UpsertedCount == actualUpsertedCount
+}
+
+func executeCollectionOperation(coll *mongo.Collection, op *operation) (bool, error) {
+	switch op.Name {
+	case "insertOne":
+		res, err := executeInsertOne(coll, op.Arguments)
+		return verifyInsertOneResult(res, op.Result), err
+	case "find":
+		cursor, err := executeFind(coll, op.Arguments)
+		return verifyCursorResult(cursor, op.Result), err
+	case "updateOne":
+		res, err := executeUpdateOne(coll, op.Arguments)
+		return verifyUpdateResult(res, op.Result), err
+	}
+	return false, errors.New("unrecognized collection operation: " + op.Name)
```

Comment:
Minor, but I'd argue for changing this to a "panic". I think the error message is not printed currently. If the test format changes, it'd be a clearer error.

Reply:
Good catch. I left a similar comment earlier about functions like `executeInsertOne` (https://github.com/mongodb-labs/drivers-atlas-testing/pull/87#discussion_r473237767). To generalize, we should only return errors once we can actually call driver functions. If there are any errors during setup for an operation, those should panic because they signal a deterministic flaw in the test runner, not a non-deterministic runtime error.

Reply:
This is an interesting discussion. Having the executor panic seems like the right thing to do here. The validator example is a bit contrived because like @iwysiu said, it uses a non-existent operation to check that the executor increments the error counter correctly. It would be much more realistic to use the executor to run an operation that is guaranteed to fail (most likely using a failPoint) to verify this behavior. I can make a quick PR to make this change if you'd rather panic here than return an error.

Reply:
@prashantmital I think that would make a lot of sense. Thanks!